lib/LongVectorLoweringPass.cpp

```diff
@@ -1546,6 +1559,213 @@ LongVectorLoweringPass::convertBuiltinCall(CallInst &VectorCall,
                                 ScalarFactory);
 }
 
+Value *LongVectorLoweringPass::convertAllBuiltinCall(
+    CallInst &CI, Type *EquivalentReturnTy, ArrayRef<Value *> EquivalentArgs) {
+  Function *Builtin = CI.getCalledFunction();
+  assert(Builtin);
+  const auto &Info = clspv::Builtins::Lookup(Builtin);
+
+  switch (Info.getType()) {
+  default:
+    return convertBuiltinCall(CI, EquivalentReturnTy, EquivalentArgs);
+  case clspv::Builtins::kShuffle:
+    return convertBuiltinShuffle(CI, EquivalentReturnTy, EquivalentArgs);
+  case clspv::Builtins::kShuffle2:
+    return convertBuiltinShuffle2(CI, EquivalentReturnTy, EquivalentArgs);
+  }
+}
+
+Value *LongVectorLoweringPass::convertBuiltinShuffle(
+    CallInst &CI, Type *EquivalentReturnTy, ArrayRef<Value *> EquivalentArgs) {
+  assert(EquivalentArgs.size() == 2);
+
+  auto Mask = EquivalentArgs[1];
+  auto MaskTy = Mask->getType();
+  unsigned MaskArity;
+  if (MaskTy->isVectorTy()) {
+    MaskArity = cast<FixedVectorType>(MaskTy)->getNumElements();
+  } else if (MaskTy->isArrayTy()) {
+    MaskArity = cast<ArrayType>(MaskTy)->getArrayNumElements();
+  } else {
+    llvm_unreachable("unexpected Type for Mask in Shuffle");
+  }
+
+  IRBuilder<> B(&CI);
+
+  Value *Src = EquivalentArgs[0];
+  auto SrcTy = Src->getType();
+  Type *ScalarTy;
+  if (SrcTy->isArrayTy()) {
+    auto SrcArrayTy = cast<ArrayType>(SrcTy);
+    ScalarTy = SrcArrayTy->getElementType();
+
+    // Because we cannot ExtractValue at a variable index from an array, we need
+    // to copy it to something where we will be able to load from a variable
+    // index
+    Value *alloca = B.CreateAlloca(SrcTy);
+    auto SrcArity = cast<ArrayType>(SrcTy)->getArrayNumElements();
+    for (uint64_t i = 0; i < SrcArity; i++) {
+      auto Val = B.CreateExtractValue(Src, i);
+      auto Gep = B.CreateGEP(alloca->getType()->getPointerElementType(), alloca,
+                             {B.getInt32(0), B.getInt32(i)});
+      B.CreateStore(Val, Gep);
+    }
+    Src = alloca;
+  } else {
+    assert(SrcTy->isVectorTy());
+    ScalarTy = SrcTy->getScalarType();
+  }
+
+  auto getScalar = [&B](Value *Vector, unsigned Index) {
+    if (Vector->getType()->isVectorTy()) {
+      return B.CreateExtractElement(Vector, Index);
+    } else {
+      assert(Vector->getType()->isArrayTy());
+      return B.CreateExtractValue(Vector, Index);
+    }
+  };
+
+  auto getScalarWithIdValue = [&B, &ScalarTy](Value *Vector, Value *Index) {
```

Comment:
The spec for shuffle says only the `ilogb(2m - 1)` (`ilogb(2m - 1) + 1` for shuffle2) bits are considered for each mask element.

`shuffle(uint2(0,1), uchar2(0, 4))` should return (0, 0), but you seem to generate an undefined value by going out of range.

There seems to missing masking here. I'm surprised CTS doesn't not catch this or am I misreading something?

Reply:
I did have a go at implementing `shuffle` a while back and noticed that the CTS indeed doesn't cover this case and also does not test the case where the mask is not a constant. I've created https://github.com/KhronosGroup/OpenCL-CTS/issues/1392.

Reply:
I also spotted https://github.com/KhronosGroup/OpenCL-Docs/issues/767 at the time and had made a note to raise it.